src/sec-updater/sec-updater.8.in
.TH SEC-UPDATER 8 "2017-08-25" "@PACKAGE_VERSION@" "strongSwan"
.
.SH "NAME"
.
sec-updater \- Extracts security update information of Linux distributions
.
.SH "SYNOPSIS"
.
.SY "sec-updater"
.OP \-\-debug level
.OP \-\-quiet
.OP \-\-security
.BI \-\-product " name"
.BI \-\-file " filename"
.YS
.
.SY "sec-updater"
.B \-h
|
.B \-\-help
.YS
.
.SH "DESCRIPTION"
.
.B sec-updater
extracts information about security updates and backports of Linux repositories
(e.g. Debian or Ubuntu). This information is used to update the package version
information stored in the strongTNC SQLite database. The "dbkg --compare-versions"
command is used to determine which package versions are affected by a given
security update.
.
.SH "OPTIONS"
.
.TP
.B "\-h, \-\-help"
Prints usage information and a short summary of the available commands.
.TP
.BI "\-v, \-\-debug " level
Set debug level, default: 1.
.TP
.B "\-q, \-\-quiet"
Disable debug output to stderr.
.TP
.B "\-s, \-\-security"
Set when parsing a distributions file with security updates.
.TP
.BI "\-p, \-\-product " name
Name of Linux version as stored in database. eg. "Ubuntu 16.04 x86_64".
.TP
.BI "\-f, \-\-file " filename
Linux package information file to parse.
.
.SH "CONFIGURATION"
.
The following parameter must be configured in strongswan.conf:
.P
 sec-updater {
    database = sqlite:///etc/pts/config.db
 }
.P
.
.SH "SEE ALSO"
.
.BR strongswan.conf (5)



NEWS
strongswan-5.6.0
----------------

- Fixed a DoS vulnerability in the gmp plugin that was caused by insufficient
  input validation when verifying RSA signatures, which requires decryption
  with the operation m^e mod n, where m is the signature, and e and n are the
  exponent and modulus of the public key.  The value m is an integer between
  0 and n-1, however, the gmp plugin did not verify this.  So if m equals n the
  calculation results in 0, in which case mpz_export() returns NULL.  This
  result wasn't handled properly causing a null-pointer dereference.
  This vulnerability has been registered as CVE-2017-11185.

- New SWIMA IMC/IMV pair implements the "draft-ietf-sacm-nea-swima-patnc"
  Internet Draft and has been demonstrated at the IETF 99 Prague Hackathon.

- The IMV database template has been adapted to achieve full compliance
  with the ISO 19770-2:2015 SWID tag standard.

- The sw-collector tool extracts software events from apt history logs
  and stores them in an SQLite database to be used by the SWIMA IMC.
  The tool can also generate SWID tags both for installed and removed
  package versions.

- The pt-tls-client can attach and use TPM 2.0 protected private keys
  via the --keyid parameter.

- libtpmtss supports Intel's TSS2 Architecture Broker and Resource
  Manager interface (tcti-tabrmd).

- The new eap-aka-3gpp plugin implements the 3GPP MILENAGE algorithms
  in software.  K (optionally concatenated with OPc) may be configured as
  binary EAP secret.

- CHILD_SA rekeying was fixed in charon-tkm and was slightly changed: The
  switch to the new outbound IPsec SA now happens via SPI on the outbound
  policy on Linux, and in case of lost rekey collisions no outbound SA/policy
  is temporarily installed for the redundant CHILD_SA.

- The new %unique-dir value for mark* settings allocates separate unique marks
  for each CHILD_SA direction (in/out).


strongswan-5.5.3
----------------

- Fixed a DoS vulnerability in the gmp plugin that was caused by insufficient
  input validation when verifying RSA signatures.  More specifically,
  mpz_powm_sec() has two requirements regarding the passed exponent and modulus
  that the plugin did not enforce, if these are not met the calculation will
  result in a floating point exception that crashes the whole process.
  This vulnerability has been registered as CVE-2017-9022.

- Fixed a DoS vulnerability in the x509 plugin that was caused because the ASN.1
  parser didn't handle ASN.1 CHOICE types properly, which could result in an
  infinite loop when parsing X.509 extensions that use such types.
  This vulnerability has been registered as CVE-2017-9023.

- The behavior during IKEv2 CHILD_SA rekeying has been changed in order to avoid
  traffic loss. The responder now only installs the new inbound SA and delays
  installing the outbound SA until it receives the DELETE for the replaced
  CHILD_SA.  Similarly, the inbound SA of the replaced CHILD_SA is not removed
  for a configurable amount of seconds (charon.delete_rekeyed_delay) after the
  DELETE has been processed to reduce the chance of dropping delayed packets.

- The code base has been ported to Apple's ARM64 iOS platform, whose calling
  conventions for variadic and regular functions are different.  This means
  assigning non-variadic functions to variadic function pointers does not work.
  To avoid this issue the enumerator_t interface has been changed and the
  signatures of the callback functions for enumerator_create_filter(), and the
  invoke_function() and find_first() methods on linked_list_t have been changed.
  The return type of find_first() also changed from status_t to bool.

- Added support for fuzzing the certificate parser provided by the default
  plugins (x509, pem, gmp etc.) on Google's OSS-Fuzz infrastructure. Several
  issues found while fuzzing these plugins were fixed.

- Two new options have been added to charon's retransmission settings:
  retransmit_limit and retransmit_jitter.  The former adds an upper limit to the
  calculated retransmission timeout, the latter randomly reduces it.

- A bug in swanctl's --load-creds command was fixed that caused unencrypted
  private keys to get unloaded if the command was called multiple times. The
  load-key VICI command now returns the key ID of the loaded key on success.

- The credential manager now enumerates local credential sets before global
  ones. This means certificates supplied by the peer will now be preferred over
  certificates with the same identity that may be locally stored (e.g. in the
  certificate cache).

- Added support for hardware offload of IPsec SAs as introduced by Linux 4.11
  for hardware that supports this.

- When building the libraries monolithically and statically the plugin
  constructors are now hard-coded in each library so the plugin code is not
  removed by the linker because it thinks none of their symbols are ever
  referenced.

- The pki tool loads the curve25519 plugin by default.


strongswan-5.5.2
----------------

- Support of Diffie-Hellman group 31 using Curve25519 for IKE as defined
  by RFC 8031.

- Support of Ed25519 digital signature algorithm for IKEv2 as defined by
  draft-ietf-ipsecme-eddsa. Ed25519-based public key pairs, X.509 certificates
  and CRLs can be generated and printed by the pki tool.

- The new "tpm" libtpmtss plugin allows to use persistent private RSA and ECDSA
  keys bound to a TPM 2.0 for both IKE and TLS authentication. Using the
  TPM 2.0 object handle as keyid parameter, the pki --pub tool can extract
  the public key from the TPM thereby replacing the aikpub2 tool. In a similar
  fashion pki --req can generate a PKCS#10 certificate request signed with
  the TPM private key.

- The pki tool gained support for generating certificates with the RFC 3779
  addrblock extension. The charon addrblock plugin now dynamically narrows
  traffic selectors based on the certificate addrblocks instead of rejecting

src/sec-updater/sec-updater.c
/*
 * Copyright (C) 2012-2017 Andreas Steffen
 * HSR Hochschule fuer Technik Rapperswil
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.  See <http://www.fsf.org/copyleft/gpl.txt>.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 */

#define _GNU_SOURCE
#include <getopt.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <syslog.h>
#include <time.h>
#include <sys/stat.h>
#include <stdlib.h>

#include <library.h>
#include <utils/debug.h>

typedef enum sec_update_state_t sec_update_state_t;

enum sec_update_state_t {
	SEC_UPDATE_STATE_BEGIN_PACKAGE,
	SEC_UPDATE_STATE_VERSION,
	SEC_UPDATE_STATE_END_PACKAGE
};

typedef struct stats_t stats_t;

struct stats_t {
	time_t release;
	int product;
	int packages;
	int new_versions;
	int updated_versions;
};

/**
 * global debug output variables
 */
static int debug_level = 1;
static bool stderr_quiet = FALSE;

/**
 * sec_updater dbg function
 */
static void sec_updater_dbg(debug_t group, level_t level, char *fmt, ...)
{
	int priority = LOG_INFO;
	char buffer[8192];
	char *current = buffer, *next;
	va_list args;

	if (level <= debug_level)
	{
		if (!stderr_quiet)
		{
			va_start(args, fmt);
			vfprintf(stderr, fmt, args);
			fprintf(stderr, "\n");
			va_end(args);
		}

		/* write in memory buffer first */
		va_start(args, fmt);
		vsnprintf(buffer, sizeof(buffer), fmt, args);
		va_end(args);

		/* do a syslog with every line */
		while (current)
		{
			next = strchr(current, '\n');
			if (next)
			{
				*(next++) = '\0';
			}
			syslog(priority, "%s\n", current);
			current = next;
		}
	}
}

/**
 * atexit handler to close everything on shutdown
 */
static void cleanup(void)
{
	closelog();
	library_deinit();
}

static void usage(void)
{
	printf("\
Usage:\n\
  sec-updater --help\n\
  sec-updater [--debug <level>] [--quiet]  [--security] --product <name> --file <filename>\n\n\
  Options:\n\
    --help             print usage information\n\
    --debug            set debug level\n\
    --quiet            suppress debug output to stderr\n\
    --security         set when parsing a file with security updates\n\
    --product <name>   name of the Linux version as stored in the database\n\
    --file <filename>  package information file to parse");
}

/**
 * Update the package database
 */
static bool update_database(database_t *db, char *package, char *version,
							bool security, stats_t *stats)
{
	int pid = 0, vid = 0, sec_flag;
	bool first = TRUE, found = FALSE;
	char *release;
	enumerator_t *e;

	/* increment package count */
	stats->packages++;

	/* check if package is already in database */
	e = db->query(db, "SELECT id FROM packages WHERE name = ?",
					  DB_TEXT, package, DB_INT);
	if (!e)
	{
		return FALSE;
	}
	if (!e->enumerate(e, &pid))
	{
		pid = 0;
	}
	e->destroy(e);

	if (!pid)
	{
		return TRUE;
	}

	/* retrieve all package versions stored in database */
	e = db->query(db,
			"SELECT id, release, security FROM versions "
			"WHERE product = ? AND package = ?",
			 DB_INT, stats->product,  DB_INT, pid, DB_INT, DB_TEXT, DB_INT);
	if (!e)
	{
		return FALSE;
	}

	while (e->enumerate(e, &vid, &release, &sec_flag))
	{
		char command[BUF_LEN];
		char found_char = ' ';
		bool update_version = FALSE;

		if (streq(version, release))
		{
			found = TRUE;
			found_char = '*';
		}
		else if (security)
		{
			 snprintf(command, BUF_LEN, "dpkg --compare-versions %s lt %s",
										 release, version);
			if (system(command) == 0)
			{
				found_char = '!';
				if (!sec_flag)
				{
					if (db->execute(db, NULL, "UPDATE versions "
						"SET security = 1 WHERE id = ?", DB_INT, vid) != 1)
					{
						DBG1(DBG_IMV, "  could not update version");
						e->destroy(e);
						return FALSE;
					}
					update_version = TRUE;
					stats->updated_versions++;
				}
			}
		}
		if (debug_level < 2 && !update_version)
		{
			continue;
		}
		if (first)
		{
			DBG1(DBG_IMV, "%s", package);
			first = FALSE;
		}
		DBG1(DBG_IMV, "  %c%s %s", found_char , sec_flag ? "s" : " ", release);
	}
	e->destroy(e);

	if (!found)
	{
		if (first)
		{
			DBG1(DBG_IMV, "%s", package);
		}
		DBG1(DBG_IMV, "  +  %s", version);

		if (db->execute(db, &vid,
			"INSERT INTO versions "
			"(package, product, release, security, time) "
			"VALUES (?, ?, ?, 0, ?)", DB_INT, pid, DB_INT, stats->product,
			DB_TEXT, version, DB_INT, stats->release) != 1)
		{
			DBG1(DBG_IMV, "  could not store version to database");
			return FALSE;
		}
		stats->new_versions++;
	}

	return TRUE;
}

/**
 * Process a package file and store updates in the database
 */
static void process_packages(char *filename, char *product, bool security)
{
	char *uri, line[BUF_LEN], *pos, *package = NULL, *version = NULL;
	sec_update_state_t state;
	enumerator_t *e;
	database_t *db;
	int pid;
	FILE *file;
	stats_t stats;
	bool success;

	/* initialize statistics */
	memset(&stats, 0x00, sizeof(stats_t));

	/* Set release date to current time */
	stats.release = time(NULL);

	/* opening package file */
	DBG1(DBG_IMV, "loading \"%s\"", filename);
	file = fopen(filename, "r");
	if (!file)
	{
		DBG1(DBG_IMV, "  could not open \"%s\"", filename);
		exit(EXIT_FAILURE);
	}

	/* connect package database */
	uri = lib->settings->get_str(lib->settings, "sec-updater.database", NULL);
	if (!uri)
	{
		DBG1(DBG_IMV, "database URI sec-updater.database not set");
		fclose(file);
		exit(EXIT_FAILURE);
	}
	db = lib->db->create(lib->db, uri);
	if (!db)
	{
		DBG1(DBG_IMV, "could not connect to database '%s'", uri);
		fclose(file);
		exit(EXIT_FAILURE);
	}

	/* check if product is already in database */
	e = db->query(db, "SELECT id FROM products WHERE name = ?",
				  DB_TEXT, product, DB_INT);
	if (e)
	{
		if (e->enumerate(e, &pid))
		{
			stats.product = pid;
		}
		e->destroy(e);
	}
	if (!stats.product)
	{
		if (db->execute(db, &pid, "INSERT INTO products (name) VALUES (?)",
						DB_TEXT, product) != 1)
		{
			DBG1(DBG_IMV, "could not store product '%s' to database",
							 product);
			fclose(file);
			db->destroy(db);
			exit(EXIT_FAILURE);
		}
		stats.product = pid;
	}

	state = SEC_UPDATE_STATE_BEGIN_PACKAGE;

	while (fgets(line, sizeof(line), file))
	{
		/* set read pointer to beginning of line */
		pos = line;

		switch (state)
		{
			case SEC_UPDATE_STATE_BEGIN_PACKAGE:
				pos = strstr(pos, "Package: ");
				if (!pos)
				{
					continue;
				}
				pos += 9;
				package = pos;
				pos = strchr(pos, '\n');
				if (pos)
				{
					package = strndup(package, pos - package);
					state = SEC_UPDATE_STATE_VERSION;
				}
				break;
			case SEC_UPDATE_STATE_VERSION:
				pos = strstr(pos, "Version: ");
				if (!pos)
				{
					continue;
				}
				pos += 9;
				version = pos;
				pos = strchr(pos, '\n');
				if (pos)
				{
					version = strndup(version, pos - version);
					state = SEC_UPDATE_STATE_END_PACKAGE;
				}
				break;
			case SEC_UPDATE_STATE_END_PACKAGE:
				if (*pos != '\n')
				{
					continue;
				}
				success = update_database(db, package, version, security, &stats);
				free(package);
				free(version);
				if (!success)
				{
					fclose(file);
					db->destroy(db);
					exit(EXIT_FAILURE);
				}
				state = SEC_UPDATE_STATE_BEGIN_PACKAGE;
		}
	}
	switch (state)
	{
		case SEC_UPDATE_STATE_END_PACKAGE:
			free(version);
			/* fall-through */
		case SEC_UPDATE_STATE_VERSION:
			free(package);
			break;
		default:
			break;
	}
	fclose(file);
	db->destroy(db);

	DBG1(DBG_IMV, "processed \"%s\": %d packages, %d new versions, "
				  "%d updated versions", filename, stats.packages,
				   stats.new_versions, stats.updated_versions);
}

static void do_args(int argc, char *argv[])
{
	char *filename = NULL, *product = NULL;
	bool security = FALSE;

	/* reinit getopt state */
	optind = 0;

	while (TRUE)
	{
		int c;

		struct option long_opts[] = {
			{ "help", no_argument, NULL, 'h' },
			{ "debug", required_argument, NULL, 'd' },
			{ "file", required_argument, NULL, 'f' },
			{ "product", required_argument, NULL, 'p' },
			{ "quiet", no_argument, NULL, 'q' },
			{ "security", no_argument, NULL, 's' },
			{ 0,0,0,0 }
		};

		c = getopt_long(argc, argv, "", long_opts, NULL);
		switch (c)
		{
			case EOF:
				break;
			case 'h':
				usage();
				exit(EXIT_SUCCESS);
			case 'd':
				debug_level = atoi(optarg);
				continue;
			case 'f':
				filename = optarg;
				continue;
			case 'p':
				product = optarg;
				continue;
			case 'q':
				stderr_quiet = TRUE;
				continue;
			case 's':
				security = TRUE;
				continue;
		}
		break;
	}

	if (filename && product)
	{
		process_packages(filename, product, security);
	}
	else
	{
		usage();
		exit(EXIT_FAILURE);
	}
}

int main(int argc, char *argv[])
{
	/* enable attest debugging hook */
	dbg = sec_updater_dbg;
	openlog("sec-updater", 0, LOG_DEBUG);

	atexit(cleanup);

	/* initialize library */
	if (!library_init(NULL, "sec-updater"))
	{
		exit(SS_RC_LIBSTRONGSWAN_INTEGRITY);
	}
	if (!lib->plugins->load(lib->plugins,
			lib->settings->get_str(lib->settings, "sec-updater.load", "sqlite")))
	{
		exit(SS_RC_INITIALIZATION_FAILED);
	}
	do_args(argc, argv);

	exit(EXIT_SUCCESS);
}

